lib/vdsm/network/nmstate/bridge_opts.py

```diff
@@ -0,0 +1,136 @@
+        if opts_str:
+            opts = opts_str.strip(self._delim).split(self._delim)
+            return dict(opt.split(self._assign_op, 1) for opt in opts)
+        return {}
+
+
+_BRIDGE_OPTS = {
+    'mac-ageing-time': int,
+    'group-forward-mask': int,
+    'hash-max': int,
+    'multicast-snooping': bool,
+    'multicast-router': int,
+    'multicast-last-member-count': int,
+    'multicast-last-member-interval': int,
+    'multicast-membership-interval': int,
+    'multicast-querier': bool,
+    'multicast-querier-interval': int,
+    'multicast-query-use-ifaddr': bool,
+    'multicast-query-interval': int,
+    'multicast-query-response-interval': int,
+    'multicast-startup-query-count': int,
+    'multicast-startup-query-interval': int,
+}
+
+
+_STP_OPTS = {
+    'enabled': bool,
+    'forward-delay': 'sec_to_hundredth_sec',
+    'hello-time': 'sec_to_hundredth_sec',
+    'max-age': 'sec_to_hundredth_sec',
+    'priority': int,
+}
+
+
+class BridgeOptsParser(OptStringParser):
+    def __init__(self, opts_by_type: {}):
+        super(BridgeOptsParser, self).__init__()
+        self._opts_by_type = opts_by_type
+        self._not_found = []
+
+    def parse(self, opts: str) -> {}:
+        """
+        param: opts: vdsm.network.nmstate.bridge_util.NetworkConfig bridge opts
+        """
+        str_dict = super().parse(opts.replace('_', '-')) if opts else {}
+        opts_dict = {}
+        for k, v in str_dict.items():
+            try:
+                opts_dict[k] = self._convert_value(k, v)
+            except KeyError:
+                self._not_found.append(k)
+        return opts_dict
+
+    @property
+    def not_found(self):
+        return self._not_found
+
+    def _convert_value(self, key, value):
+        conversion_func = self._opts_by_type[key]
+        if conversion_func == bool:
+            conversion_func = self._str_to_bool
+        elif conversion_func == 'sec_to_hundredth_sec':
+            conversion_func = self._multiply_by_hundred
+        return conversion_func(value)
+
+    @staticmethod
+    def _str_to_bool(str_value):
+        return str_value.lower() not in ['false', '0']
+
+    @staticmethod
+    def _multiply_by_hundred(str_value):
+        return int(int(str_value) / 100)
+
+
+class BridgeOptsBuilder:
+    def __init__(self):
+        self._bridge_opts_parser = BridgeOptsParser(_BRIDGE_OPTS)
+        self._stp_opts_parser = BridgeOptsParser(_STP_OPTS)
+
+    def parse(self, opts: str, stp=False) -> {}:
+        """
+        param: opts: vdsm.network.nmstate.bridge_util.NetworkConfig bridge opts
+        """
+        bridge_opts = self._bridge_opts_parser.parse(opts)
+        bridge_opts[LinuxBridge.STP_SUBTREE] = {LinuxBridge.STP.ENABLED: stp}
+        stp_opts = self._stp_opts_parser.parse(opts)
+        self._check_for_unsupported_opts()
+        bridge_opts[LinuxBridge.STP_SUBTREE].update(stp_opts)
+        return bridge_opts
+
+    def _check_for_unsupported_opts(self):
```

Comment:
I am not sure what is the purpose of this function. Is there a reason why we cannot raise right away in the `parse`?

Reply:
rearranged the flow to traverse the options schema instead of the provided options string. Hope its neater now